### PR TITLE
Sync ``libasr`` with LPython

### PR DESCRIPTION
**Pull request by czgdp1807**
cc: @certik @Thirumalai-Shaktivel @Shaikh-Ubaid @konradha 

Corresponding libasr PR - https://github.com/lcompilers/libasr/pull/4
Corresponding LPython PR - https://github.com/lcompilers/lpython/pull/1091

### CODE REVIEW

File: src/libasr/asdl_cpp.py [hunk 1]
```diff
@@ -1190,10 +1217,12 @@ def visitModule(self, mod):
...
         self.emit("public:")
-        self.emit(  "std::string s, indtd = \"\";", 1)
+        self.emit(  "std::string s, indtd;", 1)
         self.emit(  "bool use_colors;", 1)
-        self.emit(  "bool indent;", 1)
-        self.emit(  "int indent_level = 0, indent_spaces = 4;", 1)
+        self.emit(  "bool indent, start_line = true;", 1)
+        self.emit(  "int indent_level = 0, indent_spaces = 3, lvl = 0;", 1)
+        self.emit(  "int tmp = 0, tmp1 = 0, tmp2 = 2;", 1)
+        self.emit(  "int curly[2000], round[2000];", 1)
```
Comment (Thirumalai-Shaktivel) [inline src/libasr/asdl_cpp.py:1225]: I submitted a PR to lfortran which altered the `--indent` handling, but I think it was not updated in LPython. So Can you please revert these changes?

Comment (czgdp1807) [reply src/libasr/asdl_cpp.py:1225]: Feel free to push commits here. :-)

File: src/libasr/compiler_tester/tester.py [hunk 2]
```diff
@@ -348,7 +340,7 @@ def tester_main(compiler, single_test):
     verbose = args.verbose
     no_llvm = args.no_llvm
 
-    # So that the tests find the `lfortran` executable
+    # So that the tests find the `lpython` executable
```
Comment (Thirumalai-Shaktivel) [inline src/libasr/compiler_tester/tester.py:343]: ..

File: src/libasr/compiler_tester/tester.py [hunk 3]
```diff
@@ -162,9 +161,6 @@ def run(basename: str, cmd: Union[pathlib.Path, str],
     if infile and not os.path.exists(infile):
         raise RunException("The input file does not exist")
     outfile = os.path.join(out_dir, basename + "." + "out")
-
-    infile=infile.replace("\\\\","\\").replace("\\","/")
```
Comment (Thirumalai-Shaktivel) [inline src/libasr/compiler_tester/tester.py:166]: This might be related to `Windows`, I don't know whether this has to be removed.

File: src/libasr/compiler_tester/tester.py [hunk 4]
```diff
@@ -126,8 +127,6 @@ def fixdir(s: bytes) -> bytes:
     local_dir = os.getcwd()
     return s.replace(local_dir.encode(), "$DIR".encode())
 
-def unl_loop_del(b):
-    return b.replace(bytes('\r\n', encoding='utf-8'), bytes('\n', encoding='utf-8'))
```
Comment (Thirumalai-Shaktivel) [inline src/libasr/compiler_tester/tester.py:130]: ..

File: src/libasr/asdl_cpp.py [hunk 5]
```diff
@@ -501,6 +501,60 @@ def visitField(self, field):
...
+         self.emit("void visit_%s(const %s_t &x) {" % (name, name), 1)
+         self.used = False
+         have_body = False
+         for field in fields[::-1]:
+             self.visitField(field)
+         if not self.used:
+             # Note: a better solution would be to change `&x` to `& /* x */`
+             # above, but we would need to change emit to return a string.
+             self.emit("if ((bool&)x) { } // Suppress unused warning", 2)
+         self.emit("}", 1)
```
Comment (Thirumalai-Shaktivel) [inline src/libasr/asdl_cpp.py:556]: I think this is repeated again

File: src/libasr/pass/pass_array_by_data.cpp [hunk 6]
```diff
@@ -226,21 +206,6 @@ class PassArrayByDataProcedureVisitor : public PassUtils::PassVisitor<PassArrayB
...
-call f(array1, array2)
-
-gets converted to,
-
-call f_array1_array2(array1, m1, n1, array2, m2, n2)
-
-As can be seen dimensional information, m1, n1 is passed along
-with array1 and similarly m2, n2 is passed along with array2.
-
-*/
```
Comment (Thirumalai-Shaktivel) [inline src/libasr/pass/pass_array_by_data.cpp:243]: I think this comment has to be restored.

File: src/libasr/pass/pass_array_by_data.cpp [hunk 7]
```diff
@@ -289,19 +254,6 @@ class ReplaceSubroutineCallsVisitor : public PassUtils::PassVisitor<ReplaceSubro
...
-as arguments to function call expressions having dimensional information of
-array arguments passed along. See example below,
-
-sum = f(array) + g(array)
-
-gets converted to,
-
-sum = f_array(array, m, n) + g_array(array, m, n)
-
-*/
```
Comment (Thirumalai-Shaktivel) [inline src/libasr/pass/pass_array_by_data.cpp:304]: ..

File: src/libasr/pass/pass_array_by_data.cpp [hunk 8]
```diff
@@ -351,13 +303,6 @@ class ReplaceFunctionCalls: public ASR::BaseExprReplacer<ReplaceFunctionCalls> {
 
 };
 
-/*
-
-The following visitor calls the above replacer i.e., ReplaceFunctionCalls
-on expressions present in ASR so that FunctionCall get replaced everywhere
-and we don't end up with false positives.
-
-*/
```
Comment (Thirumalai-Shaktivel) [inline src/libasr/pass/pass_array_by_data.cpp:360]: ..

File: src/libasr/pass/pass_array_by_data.cpp [hunk 9]
```diff
@@ -376,16 +321,6 @@ class ReplaceFunctionCallsVisitor : public ASR::CallReplacerOnExpressionsVisitor
...
-/*
-
-Since the above visitors have replaced procedure and calls to those procedures
-with arrays as arguments, we don't need the original ones anymore. So we remove
-them from the ASR. The reason to do this is that some backends like WASM don't
-have support for accepting and returning arrays via array descriptors. Therefore,
-they cannot generate code for such functions. To avoid backends like WASM from failing
-we remove those functions by implementing and calling the following visitor.
-
-*/
```
Comment (Thirumalai-Shaktivel) [inline src/libasr/pass/pass_array_by_data.cpp:388]: ..

**Comment by Thirumalai-Shaktivel** [comment]
Apart from the above changes, this PR LGTM!
Thanks for updating the libasr.

**Comment by czgdp1807** [comment]
@certik @Thirumalai-Shaktivel @Shaikh-Ubaid @konradha @Pranavchiku Please go through the changes of this diff and let me know if any of your changes are absent or you think something is incorrect (like you made a bug fix recently but its reverted here, and stuff like that). I will wait for a couple of hours before syncing LPython with LFortran's libasr.

**Comment by Thirumalai-Shaktivel** [comment]
LGTM!

**Comment by Shaikh-Ubaid** [comment]
@czgdp1807 I added commits to remove the `StatementsFirstWalkVisitorVisitor` and the pass `arr_dims_propagate`. `StatementsFirstWalkVisitorVisitor` was used by the pass `arr_dims_propagate` and `arr_dims_propogate` was superseded by `pass_array_by_data()`.

**Comment by Shaikh-Ubaid** [comment]
It looks good to me. Thank you.

**Review by Smit-create — APPROVED** [review]

**Comment by Pranavchiku** [comment]
looks good to me :)

**Comment by konradha** [comment]
AFAICT this looks good, I'm having a closer look now for the changes I have some knowledge of.

**Comment by certik** [comment]
I think so, I assume this is just copied `libasr`, and a few tiny changes in LFortran frontend.

**Comment by czgdp1807** [comment]
> I think so, I assume this is just copied `libasr`, and a few tiny changes in LFortran frontend.

Yes. I am going to sync LPython's libasr from LFortran now.

**Review by konradha — APPROVED** [review]

**Comment by czgdp1807** [comment]
The reasons for following the syncing approach are listed in https://github.com/lcompilers/libasr/pull/2#issuecomment-1174829287 and @certik’s experience from symengine in https://github.com/lcompilers/libasr/pull/2#issuecomment-1175582830

**Comment by konradha** [comment]
@czgdp1807 I'll post those links the next time someone asks about it ;) Thanks a lot!

**Comment by czgdp1807** [comment]
Sure. We should document this in the wiki somewhere.
